Description: Reject filters when the block length is nonsensical
 A flaw was identified in the RAR5 archive decompression logic of the libarchive
 library. When a specially crafted RAR5 archive is processed, the decompression
 routine may enter a state where internal logic prevents forward progress. This
 condition results in an infinite loop that continuously consumes CPU resources.
 .
 This fix adds an additional sanity check to reject filters when the block
 length is larger than half of the window size, preventing the infinite loop.
Author: Tim Kientzle <kientzle@acm.org>
Origin: upstream, https://github.com/libarchive/libarchive/commit/7273d04803a1e5a482f26d8d0fbaf2b204a72168
Author: Tim Kientzle <kientzle@acm.org>
Origin: upstream, https://github.com/libarchive/libarchive/commit/7273d04803a1e5a482f26d8d0fbaf2b204a72168
Bug: https://security-tracker.debian.org/tracker/CVE-2026-4111
Bug-Debian: https://bugs.debian.org/1130753
Forwarded: not-needed
Last-Update: 2026-04-15

--- a/libarchive/archive_read_support_format_rar5.c
+++ b/libarchive/archive_read_support_format_rar5.c
@@ -2915,7 +2915,9 @@ static int parse_filter(struct archive_read* ar, const uint8_t* p) {
 	if(block_length < 4 ||
 	    block_length > 0x400000 ||
 	    filter_type > FILTER_ARM ||
-	    !is_valid_filter_block_start(rar, block_start))
+	    !is_valid_filter_block_start(rar, block_start) ||
+	    (rar->cstate.window_size > 0 &&
+	     (ssize_t)block_length > rar->cstate.window_size >> 1))
 	{
 		archive_set_error(&ar->archive, ARCHIVE_ERRNO_FILE_FORMAT,
 		    "Invalid filter encountered");
